### ISSUE TITLE
node receiving same message more than once   

### ISSUE THREAD

**Issue by nadendlaprasad** [savoirfairelinux/opendht#360]
I have noticed when we started couple of nodes. Sometimes nodes is receiving same message more than once. any idea why this is happening ?

**Comment by aberaud** [comment]
Hi,
Thanks for reporting this,
It should never happen,

Could you please describe your setup and what ops you run ?

**Comment by nadendlaprasad** [comment]
Hi , We started three instances of opendht. Once all these nodes are connected , they start exchanging the messages among each others. Couple of times we have noticed that a node is receiving the same message repeatedly . In the logs i can see `Dropping packet with high delay: 0.519905` messages couple of times. Is this causing any abnormal  behavior ?

I'm also printing the logs whenever the status of node is changing by passing callback function to setOnStatusChanged function. Following is the sample code

```
node.setOnStatusChanged([](dht::NodeStatus status, dht::NodeStatus status2) {

			if (status == dht::NodeStatus::Disconnected) {
				std::cout << "Node is disconnected to network ";
			}
			else if (status == dht::NodeStatus::Connected) {
				std::cout << "Node is connected to network ";
			}
			else if (status == dht::NodeStatus::Connecting) {
				std::cout << "Node is connecting to network... ";
			}

		});

```
This callback is getting firing quite often and logging following messages.

`Node is connecting to network... Node is connected to network. Node is connecting to network... Node is connected to network`

We are running all the three nodes in the same network. do you think this issue is occurring due to breaks in our local network or any other idea ?

Is there any ping timeouts that we can customize ?

**Comment by nadendlaprasad** [comment]
Thanks. any idea why the nodes are receiving the same message repeatedly and why status of node is keep changing from connecting to connected . We did try running the nodes in Amazon too. we are facing the same issue.

**Comment by nadendlaprasad** [comment]
Adding more information:
We are using node.listen to receive new messages and performing put operation to send the message on specific channel.

**Comment by aberaud** [comment]
Hi,
With the new listen API (introduced in 1.7 / 1.8), listen will call the callback once for a new value and once when the value expire. The boolean value (second argument of the listen callback) should be true when the the value is expired. If the callback only takes a single argument (the value) then it should only be called once.

What is the timing of apparence of duplicate values ?

**Comment by munawarb** [comment]
Hi @aberaud , We implemented an ID system so the callback can detect and drop duplicate messages. This has solved that problem.

We're upgrading to the latest version of OpenDHT and will let you know how the testing goes.

One thing I've noticed while playing with the DhtChat tool is that the node doesn't detach from a key even if the callback returns false. I modified the callback in the dhtchat.cpp to return false instead of true and the node keeps receiving messages when they are sent. According to the documentation, when a callback returns false the node will stop listening on the key. We thought this was the problem in our project since we were inadvertently returning false if the message received was from the node that sent it, but it seems like OpenDHT is ignoring the return value in the callback anyway.

In our project we're using the method you've laid out in dhtchat.cpp to send messages among nodes, using the immessage object just as you've done in dhtchat.cpp.

While running two nodes locally, I sent a chat message from one node. I got a "message publishing failed!" message. Yet when I connected the second node and had it join the room the first one was listening on and sending to, the second node received the message that the first node said had failed to publish. This is using the dhtchat tool directly and not our project.

**Comment by aberaud** [comment]
Hi and thanks for your feedback,

Could you please try with latest master ? 0c1e7d27635ed27e1dbf9f37619f26bf94fb819b

It should fix the latest issue you mentioned and greatly speedup the local network bootstraping time.

For the other issues (duplicate value, not respecting the return value) I tried with master and coudn't reproduce. What version where you using before upgrading ?

**Comment by munawarb** [comment]
Hi @aberaud , Thanks for your continued responsiveness in helping with our issue!

@nadendlaprasad  pulled the latest Master branch and compiled it so we're now using the latest version. I'm also linking dhtChat against the latest lib. I'm still seeing it not honor the return value. The code below should prevent the sender from seeing any messages, and the receiver to see only one message. This is because the handler returns `false`. However, I'm seeing that all messages are continuously received by all parties

Here's my modification:
```
					dht.listen<dht::ImMessage>(room, [&](dht::ImMessage&& msg) {
						if (msg.from != myid)
							std::cout << msg.from.toString() << " at " << printTime(msg.date)
							<< " (took " << print_dt(std::chrono::system_clock::now() - std::chrono::system_clock::from_time_t(msg.date))
							<< "s) " << (msg.to == myid ? "ENCRYPTED " : "") << ": " << msg.id << " - " << msg.msg << std::endl;
						return false;
					});
```

As you can see, I've changed `return true` to `return false`, which should stop any messages from being received by the sender (because they do receive their own message as expected,) but still all chat messages are received by everyone.

Are we pulling from the correct branch here or is there another repo we should get your updates from?

We are noticing some improvements in reliability of message delivery, and will get back to you when we've tested this more.

**Comment by munawarb** [comment]
Hi @aberaud , While testing, we noticed that if a node sends a message at the same time the node is trying to receive a message, the received message is dropped. Also, we had three nodes connected using the chat tool and we were rapid-fire sending messages from two of the nodes. We noticed that the third node wasn't reliably receiving all the messages sent the the two nodes.

We're also getting messages now and then that say "dropping packet with high delay."

**Comment by aberaud** [comment]
Hi,
I'm investigating the return value not honored issue as well as reliability issues,

About "dropping packet with high delay.", this usually happens when the node is limited by the speed of crypto operations (RSA encryption/decryption, using 4094 bits keys by default). To avoid packets to be treated too late, received packets that are late for treatment by more than 0.5s are dropped. This behavior might be adjusted depending on your hardware, needs, and network conditions.

**Comment by munawarb** [comment]
Hi @aberaud  , Thanks for this information. This would definitely be an issue in our case since we are using `putSigned`. Is there a way to resolve this, by using shorter keypairs perhaps? I don't see anywhere in `dht::crypto` to adjust the bitness of the generated keypair.

Also, thanks for your quick attention to the rest of these issues. Our project is heavily dependent on OpenDHT because it quite literally does everything we need, especially now that it has a listen API. Do you know approximately when you'll have these issues resolved?

**Comment by aberaud** [comment]
Hi,
Latest master commit should greatly improve performance for your use case
Still trying to reproduce the return false issue

**Comment by munawarb** [comment]
Awesome, thanks for the quick update! We'll let you know how it goes.

**Comment by nadendlaprasad** [comment]
Hi, we pulled the latest branch and compiled it, we are still observing that the messages are dropping. we are also facing new issue which is , if we try to fetch the data from from dht using get method future.get() is going to infinite state if there is no such key in the dht . Earlier we used to get empty string from the dht.

**Comment by aberaud** [comment]
Hi,
I added a unit test for the value duplication/return false/token for empty hash issues:
https://github.com/savoirfairelinux/opendht/commit/56773baa9882c6fca6579066b9060bb2b9adba07

**Comment by nadendlaprasad** [comment]
Thank you. I have written a small program to simulate the issue 

```
#include <iostream>
#include <opendht.h>

int main()
{
	dht::DhtRunner node;

	// Launch a dht node on a new thread, using a
	// generated RSA key pair, and listen on port 4222.
	node.run(4222, dht::crypto::generateIdentity(), true);

	// put data into dht
	std::future<void> result(std::async([&node] {
		int i = 0;
		while (i <5000) {
			i++;
			std::thread t([&node, &i] {
				node.put(dht::InfoHash::get("unique_key" + i),"value: "+i);				
			});
			t.detach();
		}
	}));
	// get data from the dht
	int i = 0;
		while (i <5000) {
			i++;
			std::thread t([&node, &i] {
				std::future<std::vector<std::string>> output = node.get<std::string>(dht::InfoHash::get("unique_key" + i));
				std::vector<std::string> result = output.get();
				std::cout << (!result.empty() ? result.back() : " " + std::to_string(i));
			});
			t.detach();
		}
	
	// wait for dht threads to end
	node.join();
	return 0;
}
```

In the above program you can observe after certain calls `output.get()` which is nothing but `future.get() ` will never responds back

I am testing this program in windows operating system and compiling using vc++ 2017.we are trying to achieve 2k operations in a second.

Hope this program helps you to find out the root cause.

**Comment by aberaud** [comment]
@nadendlaprasad Thanks,
multiple remarks:

- `put()` is always non-blocking and async in OpenDHT, so there is no need to run it in a separate thread
- `i` is captured by reference (only a pointer to i is copied in the lambdas) so when the lambda executes, i might have a different value than during the loop !
- make sure to have waited on your futures or destroyed them before calling dht.join()

There is a modified version:

```c++
int main (int argc, char *argv[])
{
    dht::DhtRunner node;
    constexpr unsigned N = 5000;
    node.run(4222, dht::crypto::generateIdentity(), true);

    std::vector<std::future<std::vector<std::string>>> futures {};
    std::vector<std::vector<std::string>> results {};
    futures.reserve(N);
    results.reserve(N);

    auto start = std::chrono::steady_clock::now();

    // put data into dht
    for (int i=0; i<N; i++)
        node.put(dht::InfoHash::get(std::string("unique_key") + std::to_string(i)), std::string("value: ")+std::to_string(i));

    // get data from the dht
    for (int i=0; i<N; i++)
        futures.emplace_back(node.get<std::string>(dht::InfoHash::get(std::string("unique_key") + std::to_string(i))));
    for (auto& f : futures)
        results.emplace_back(f.get());
    futures.clear();

    auto delay = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start).count();
    auto rate = N * 1000 / delay;
    std::cout << "took " << delay << " ms, " << rate << " put+get per second" << std::endl;

    // wait for dht threads to end
    node.join();
    return 0;
}
```

Output on my machine:
`took 1457 ms, 3431 put+get per second`

**Comment by nadendlaprasad** [comment]
@aberaud Thanks for your response.

We have written the program intentionally to simulate the issue that we are facing.

The issue that we are facing is , when we are making multiple put and get calls in parallel to DHT, If the key is not found inside the DHT, future.get() is never responding back. This is only occurring when the through put is very high on DHT, basically we are making high volume of PUT and GET calls to DHT here  in parallel. To simulate this issue we are executing put operations in an asynchronous thread so that we can make the GET calls also in parallel. In this case it is okay if the PUT operation is not occurred on a specific key before we make GET call to the DHT. But the problem here is that DHT is not returning any response at all if the key is not found inside DHT when the through put is high on DHT where as it supposed to return empty.

Hope I could able to make you understand the problem.

**Comment by aberaud** [comment]
@nadendlaprasad Thanks,
I think I understand,
I profiled execution and found a behavior that was crippling performance in bad ways when using a node alone (disconnected from the network).

The latest master should fix the issue. Could you please try it ?

**Comment by nadendlaprasad** [comment]
@aberaud  Thank you very much for your support.

We pulled the latest changes and tested with the program that we have shared with you. we are not facing hanging issue with `future.get()` any more while fetching data from DHT . However, when we use DoneCallback with Put operation it is always returning false even though it is successfully putting the data inside DHT.  we confirmed this by fetching the data back from DHT once we made the Put operation. 

Here is the sample code. we have tested this with master branch version.

```
node.put(dht::InfoHash::get("mykey"), dht::Value("value to be stored in dht"), [](bool success) {
		if(!success)
		std::cout << "DHT put failed : "<< std::endl;
	});
```
we did try with the following version also, facing the same issue

```
const char* my_data = "Value to be stored inside dht";
node.put(dht::InfoHash::get("mykey"),  dht::Value((const uint8_t*)my_data, std::strlen(my_data)), [](bool success) {
		if(!success)
		std::cout << "DHT put failed : "<< std::endl;
	});
```

do you see anything wrong with the above sample code ?

**Comment by aberaud** [comment]
Hi,
The reason is that the DHT is disconnected so donecb would always return false even if the data is properly stored on the local node.
Callback would return true when the node is connected to the network.

**Comment by aberaud** [comment]
Do you still face the value duplication issue ? Otherwise can I close the issue ?

**Comment by munawarb** [comment]
Hi @aberaud , Thanks so much for your help through this thread; we really appreciate the promptness of responses and fixes. We worked around the value duplication issue by attaching message GUIDs to all messages we send over channels. I'm not sure if we're still having problems, but if the testers report further issues we'll open up a new thread, so you can go ahead and close this one for now. Thanks again.
